### PR TITLE
Implement full text search on book title and author

### LINKED ISSUE #20: Research fuzzy search with Firestore
- Search keywords to match titles and authors of books
- Try filtering for only not accepted/borrowed books

### LINKED ISSUE #109: Implement full text search using Algolia
Implement #20

### PR DESCRIPTION
**Pull request by armiantos**
# Summary
- Pressing enter after entering keywords in the search bar will update list of books
- Keywords are space separated and will be searched on both title and author via Algolia
- There is some data redundancy (storing IDs, titles, and authors in Algolia)
- Most of the code was from [Algolia Android - Getting started](https://www.algolia.com/doc/guides/getting-started/quick-start/tutorials/quick-start-with-the-api-client/android/?language=android)
- On adding a new book, a new entry is also added to Algolia
- If a collection update happens after search, repository will automatically re-search the last query

# Screenshots
https://photos.app.goo.gl/4EdjrscdvC5MA2gB6

# TODO
- Add test cases
- Handle multi paged results from Algolia
- If there was any new books added from 6 November, we need to add those indices to Algolia or clear the `books` collection

Closes #109 
Closes #20

### CODE REVIEW

File: app/src/main/java/com/example/unlibrary/library/LibraryRepository.java [hunk 1]
```diff
@@ -179,4 +198,29 @@ public void setFilter(FilterMap filter) {
...
+
+    /**
+     * Adds book to Algolia search index so it can be searched in {@link com.example.unlibrary.exchange.ExchangeFragment}.
+     * Book's id, title, and author are assumed to be non null.
+     *
+     * @param book new book to add to search index
+     */
+    public void addAlgoliaIndex(Book book) {
+        Index index = mAlgoliaClient.getIndex(ALGOLIA_INDEX_NAME);
+        try {
```
Comment (armiantos) [inline app/src/main/java/com/example/unlibrary/library/LibraryRepository.java:210]: Try-catch is to catch `JSONObject.put`. I've never seen this happen though.

Comment (armiantos) [reply app/src/main/java/com/example/unlibrary/library/LibraryRepository.java:210]: Algolia is REST based so it uses JSON to transfer data. Refer to [Getting Started](https://www.algolia.com/doc/guides/getting-started/quick-start/tutorials/quick-start-with-the-api-client/android/?language=android) for more information.

**Review by armiantos — COMMENTED** [review]
Some guidelines to look for while reviewing

File: app/src/main/java/com/example/unlibrary/exchange/ExchangeRepository.java [hunk 2]
```diff
@@ -123,5 +136,48 @@ public void createRequest(Request request, OnSuccessListener<DocumentReference>
...
+        Query query = new Query(keywords).setAttributesToRetrieve(ALGOLIA_ID_FIELD);
+
+        index.searchAsync(query, (content, error) -> {
+            if (error != null) {
+                Log.e(TAG, "Algolia error: ", error);
+                return;
+            }
+
+            // Get results array
+            JSONArray hits = content.optJSONArray("hits");
```
Comment (armiantos) [inline app/src/main/java/com/example/unlibrary/exchange/ExchangeRepository.java:159]: For some reason the search results is stored within a JSON field called hits(?) [JS reference](https://firebase.google.com/docs/firestore/solutions/search)

**Review by armiantos — COMMENTED** [review]
More guidelines to look for while reviewing

**Review by calebschoepp — DISMISSED** [review]
Great work @armiantos . Really clean code. LGTM!

File: app/src/main/java/com/example/unlibrary/exchange/ExchangeFragment.java [hunk 3]
```diff
@@ -55,6 +55,13 @@ public void onCreate(@Nullable Bundle savedInstanceState) {
     public View onCreateView(LayoutInflater inflater, ViewGroup container, Bundle savedInstanceState) {
         mBinding = FragmentExchangeBinding.inflate(inflater, container, false);
         mBinding.setLifecycleOwner(getViewLifecycleOwner());
+        mBinding.setViewModel(mViewModel);
+
+        // Add listener to enter key in search bar
+        mBinding.searchExchangeBook.setOnEditorActionListener((textView, i, keyEvent) -> {
+            mViewModel.search(textView.getText().toString());
+            return true;
+        });
```
[the remark this answers is not in the document]
Comment (armiantos) [reply app/src/main/java/com/example/unlibrary/exchange/ExchangeFragment.java:64]: Thanks! It does it on pressing enter. Although the docs mention that an enter can be detected if `keyEvent` is not null. It's always null for me and this is only triggered on Enter.

**Review by calebschoepp — APPROVED** [review]
LGTM. Great work @armiantos

**Comment by armiantos** [comment]
Thanks! @calebschoepp

**Pull request merged**